[3d] Ambient Occlusion (reworked)

wonder-sk opened the pull request:
Adds ambient occlusion effect to 3D views. This effect can nicely help with depth perception of the data in 3D - see screenshots below. This is a post-processing effect and it can be even combined with eye dome lighting (EDL) for quite interesting results.

This PR is a replacement for #49144 - mostly keeping @NEDJIMAbelgacem's infrastructure work, but changing the actual SSAO shader implementation. It should fix the issues pointed out by @nyalldawson in the previous PR.

Initially we based the code on CloudCompare's implementation, however that did not work too well for us for a couple of reasons:
- the code does not deal well with perspective projection, causing incorrect shading (CC uses orthographic projection)
- there was no range check, so we would be getting false ambient occlusion on larger depth discontinuities (silhouttes)
- banding artifacts as the sampling kernel was not getting rotated
- parameters (shading radius, distance attenuation) that are difficult to understand

The new implementation is based on John Chapman's tutorial and LearnOpenGL page (derived from the original tutorial):
https://john-chapman-graphics.blogspot.com/2013/01/ssao-tutorial.html
https://learnopengl.com/Advanced-Lighting/SSAO

The general approach of the SSAO is the following:
- for each pixel, we pick a couple of random points nearby (64 samples currently) and check with the depth buffer whether they are visible from the camera or not. The nearby points that are occluded contribute to darkening of the pixel, this is saved to a texture
- in the next rendering pass, we blur the texture using 4x4 box. This is because in the first step we use 4x4 random noise pattern and it leaves a noticeable noise pattern on the screen. This pass gets rid of that noise
- in the post-processing step, the blurred texture is blended with the rendered scene

There are few differences to J.C.'s tutorial and LearnOpenGL page:
- the approches above use normal maps (a texture with a normal vector for each pixel), but we don't because we also want to support point clouds that do not have normals (at least not by default)
- we use full sphere for sampling instead of hemisphere (which is possible when you have normals), so maybe we are getting a bit lower quality / performance
- LearnOpenGL also uses a texture with positions of all pixels - we only use depth map to get the original positions (like JC's original code does)

For anyone interested to play further with SSAO, I have made a prototype implementation in Qt3D here (with no QGIS code): https://github.com/wonder-sk/qt3d-experiments/tree/master/ssao

## GUI

There is a new item in the configuration menu of 3D views:

![ssao-config-1](https://user-images.githubusercontent.com/193367/183769526-b2af8508-682e-4504-a525-a2816b740d69.png)

And there are extra parameters in the advanced section of the configuration dialog:
- radius - how far we will reach to calculate ambient occlusion
- intensity - how strong the effect should be (higher values make things darker)
- occlusion threshold - how many neighboring points need to be occluded for the effect to appear (lower values than 50% will make the output darker, but possibly providing greater range of occlusion)

![ssao-config](https://user-images.githubusercontent.com/193367/187551420-eadffe5d-8890-4bdd-b999-d5d58d925389.png)

## Screenshots

**Trencin castle** (SSAO parameters: radius = 50 meters, intensity = 0.7)

| SSAO off        | SSAO on |
| ------------- | ------------- |
| ![tn](https://user-images.githubusercontent.com/193367/183767713-d5c3b68b-47de-4a66-b23f-b02b638991df.png)      | ![tn-ssao](https://user-images.githubusercontent.com/193367/183767737-d98355a9-dbda-444e-8e5b-9cad358168ba.png) |

| EDL + SSAO off        | EDL + SSAO on |
| ------------- | ------------- |
| ![tn-edl](https://user-images.githubusercontent.com/193367/183767819-9afe9dc4-2a3e-493a-b2c4-627f1a905320.png) | ![tn-ssao-edl](https://user-images.githubusercontent.com/193367/183767870-10ea8d08-fc30-4ba3-bade-b32dfdede73c.png) |


**Trencin trees** (SSAO parameters: radius = 50 meters, intensity = 1.2)

| SSAO off        | SSAO on |
| ------------- | ------------- |
| ![trees](https://user-images.githubusercontent.com/193367/183768343-4d383a95-9a0f-4c29-b19d-636be1adee63.png) | ![trees-ssao](https://user-images.githubusercontent.com/193367/183768363-87718bdd-bcb0-4650-aa3c-8aed2cfe4e03.png) |

| EDL + SSAO off        | EDL + SSAO on |
| ------------- | ------------- |
| ![trees-edl](https://user-images.githubusercontent.com/193367/183768386-2da8b8ef-f9f1-46fe-a216-72405ef05bc3.png) | ![trees-ssao-edl](https://user-images.githubusercontent.com/193367/183768401-1b2e81f4-8c0c-4a17-adcd-08b0723522b4.png) |

**Prague buildings** (SSAO parameters: radius = 20 meters, intensity = 0.5)

| SSAO off        | SSAO on |
| ------------- | ------------- |
| ![buildings](https://user-images.githubusercontent.com/193367/183768790-cc50d94c-c6d4-451e-928a-0b634dd9f90b.png) | ![buildings-ssao](https://user-images.githubusercontent.com/193367/183768804-4cf96cd4-f112-4d3a-b524-efcddc6865ad.png) |

NEDJIMAbelgacem committed:
Initial implementation

NEDJIMAbelgacem committed:
- Add disabling of blur pass
- Change default parameter for shading factor
- fix some tests

NEDJIMAbelgacem committed:
- Implement bilateral filtering
- Optimize textures
- Fix layout test

NEDJIMAbelgacem committed:
rename SSAO to ambient occlusion
fix ssao settings widget margin

NEDJIMAbelgacem committed:
Remove unused variable

NEDJIMAbelgacem committed:
fix naming and add tooltips

NEDJIMAbelgacem committed:
- Refactor quad entities
- Address Stefanos's suggestions

wonder-sk committed:
Rework SSAO implementation

Previously we based the code on CloudCompare's implementation,
however that did not work too well for us for a couple of reasons:
- the code does not deal well with perspective projection, causing incorrect shading (CC uses orthographic projection)
- there was no range check, so we would be getting false ambient occlusion on larger depth discontinuities (silhouttes)
- banding artifacts as the sampling kernel was not getting rotated
- parameters (shading radius, distance attenuation) that are difficult to understand

The new implementation is based on John Chapman's tutorial and LearnOpenGL page (derived from the original tutorial):
https://john-chapman-graphics.blogspot.com/2013/01/ssao-tutorial.html
https://learnopengl.com/Advanced-Lighting/SSAO

The general approach of the SSAO is the following:
- for each pixel, we pick a couple of random points nearby (64 samples currently)
  and check with the depth buffer whether they are visible from the camera or not.
  The nearby points that are occluded contribute to darkening of the pixel,
  this is saved to a texture
- in the next rendering pass, we blur the texture using 4x4 box. This is because
  in the first step we use 4x4 random noise pattern and it leaves a noticeable noise
  pattern on the screen. This pass gets rid of that noise
- in the post-processing step, the blurred texture is blended with the rendered scene

There are few differences to J.C.'s tutorial and LearnOpenGL page:
- the approches above use normal maps (a texture with a normal vector for each pixel),
  but we don't because we also want to support point clouds that do not have normals
  (at least not by default)
- we use full sphere for sampling instead of hemisphere (which is possible when you
  have normals), so maybe we are getting a bit lower quality / performance
- LearnOpenGL also uses a texture with positions of all pixels - we only use depth map
  to get the original positions (like JC's original code does)

wonder-sk committed:
Clean up ssao parameters and GUI, add intensity parameter

nyalldawson commented:
Screenshots look beautiful, lovely work @wonder-sk ! I'll give this a test run asap.

One remaining question I have regarding the screenshots (which also applies to the shadow/eye dome effect)... is there anyway we can adjust the gamma/brightness of these effects to avoid the overall darkening of the scene which happens when they are enabled?

Eg. ideally in this area where there's no occlusion happening the brightness of the scene would remain unchanged:

![image](https://user-images.githubusercontent.com/1829991/183772069-a74f2556-dd58-4b68-9fa0-20aaff00ec56.png)

wonder-sk commented:
Yeah that darkening has been bothering be as well a bit. There's also a related possibly unwanted effect where outer edges look brighter than the rest of walls. As far as I understand, it is a cost of sampling neighbors from the full sphere (rather than just using correctly oriented hemisphere), so on a planar surface roughly half of the samples end up occluded, causing the darkening.

I have tried to fix it by adding the assumption that half of the samples get occluded - in `ssao_factor_render.frag` in `ssao()` just before the return ([line 57](https://github.com/qgis/QGIS/pull/49702/files#diff-22c390d2897aa27e1ddc74d1cd77b2c8a1c08bdfe9dc480f6e9cffa75ed09981R57)), something like this:
```glsl
occlusion = max(0, occlusion-32);
```
and modifying the `return` line below to divide by 32 instead of 64.

While that helps with these planar surfaces (no darkening, edges not getting brighter), the SSAO effect on point clouds started to look very dull... which is understandable I guess, there's much more variety of occlusion to be expected there, and we have simply cut out half of the output range. Maybe it could be treated by another attribute to cater for both kinds of scenarios.

wonder-sk commented:
@nyalldawson to the more general question about gamma/brightness adjustment of both eye dome lighting and ambient occlusion - that is something that could be definitely added, both of these effects simply produce value in range [0...1] for each pixel, which we then use to multiply the original color in the final step.

Actually the "intensity" parameter is essentially gamma correction for ambient occlusion, as it is only applied at the end of SSAO calculation like this: `output = pow(value, intensity)`

DelazJ commented:
@wonder-sk Thanks for the what, why and how. Much appreciated.

DelazJ commented:
On another topic: the advanced tab is getting bigger.. Maybe should we see whether some related options could be grouped in a dedicated tab?

wonder-sk commented:
> On another topic: the advanced tab is getting bigger.. Maybe should we see whether some related options could be grouped in a dedicated tab?

@DelazJ definitely... I can think of two new tabs:
- post-processing / effects - for configuration of EDL, SSAO or other effects in the future
- for developers / debugging - all the bits that normally people should not need (bounding boxes, map tile info, camera's view/rotation center, shadow map, depth map)

It would be best to leave such changes for a separate PR though...

wonder-sk committed:
Add missing Q_OBJECT macro

wonder-sk committed:
Add more missing Q_OBJECT macros

NEDJIMAbelgacem commented on src/3d/shaders/ssao_factor_render.frag:
```diff
@@ -0,0 +1,76 @@
+float ssao(vec3 originPos, vec3 noise)
+{
+    float occlusion = 0.0;
+    for (int i = 0; i < kernelSize; ++i)
```
```
    //	project pixel position on screen:
    vec4 originPosScreen = origProjMatrix * vec4(originPos, 1.0);
    originPosScreen.xy = originPosScreen.xy / originPosScreen.w  * 0.5 + 0.5;

    //	get sample depth:
    float depth = texture(depthTexture, originPosScreen.xy).r;
    if ( depth == 1 )
      return 1.0f;
```

NEDJIMAbelgacem replied:
This added code will avoid making the scene darker where nothing is rendered

wonder-sk replied:
I think Nyall was concerned not only about the background (where depth==1), but also the foreground (terrain, buildings etc.), as highlighted in his screenshot, so I am afraid this is not going to help...

uclaros replied:
Skipping the background definitely helps though, I believe this check should be included.
Also, lowering the default intensity to `0.5` instead of `1` gives a nicer looking result imho. The darkening is still there, but less... dark :)

NEDJIMAbelgacem reviewed (changes requested):
```
    //	project pixel position on screen:
    vec4 originPosScreen = origProjMatrix * vec4(originPos, 1.0);
    originPosScreen.xy = originPosScreen.xy / originPosScreen.w  * 0.5 + 0.5;

    //	get sample depth:
    float depth = texture(depthTexture, originPosScreen.xy).r;
    if ( depth == 1 )
      return 1.0f;
```

nyalldawson commented:
(triggering a rebuild to run through new qt 6 workflows...)

uclaros reviewed (commented):
Would it be possible to split the ssao effect into a both darkening and lightening effect? So instead of darkening all pixels depending on the calculated ssao value [0.-1.] we would use values [0.-.5] to darken by [.5-1.] and use values [.5-1.] to lighten by [0.-.5] ?
This could maybe help to avoid having the scene appear too dark.

wonder-sk committed:
Add occlusion threshold parameter to control the darkening

With the default threshold of 50%, pixels only start to get darker
when more than half of the neighborhood samples are occluded. That
means flat surfaces should not get any darker. (What we had previously
is an equivalent of having threshold set at 0%)

The downside is that with increased threshold, more subtle occlusions get lost.

wonder-sk committed:
Review from Stefanos

wonder-sk committed:
More review and better defaults

wonder-sk commented:
I have added a new parameter "occlusion threshold" that can control the amount of "darkening" - in the default configuration (50%) there should be no overall image darkening.

Some samples of how the threshold affects rendering: with low values there is perceivable "darkening" but greatest visual range of the occlusion effect, with higher values there is less to no darkening, but the ambient occlusion is less pronounced.

| 0%        | 25% |
| ------------- | ------------- |
| ![ssao-thr-0](https://user-images.githubusercontent.com/193367/187552026-870dd13c-6e7b-4d28-843d-b1cba4954c1b.png) | ![ssao-thr-25](https://user-images.githubusercontent.com/193367/187552045-c9e12f8a-313f-4113-9415-96c1d445ba82.png)  |

| 50%        | 75% |
| ------------- | ------------- |
| ![ssao-thr-50](https://user-images.githubusercontent.com/193367/187552062-3c06692c-1050-43f9-a00d-f47867350b1f.png) | ![ssao-thr-75](https://user-images.githubusercontent.com/193367/187552076-55c59d81-a1d3-4605-a327-5a61c8a864c2.png) |

nyalldawson commented:
Hey @wonder-sk I've just tested the latest version and all looks great!!

uclaros commented on src/app/3d/qgsambientocclusionsettingswidget.cpp:
```diff
@@ -23,6 +23,10 @@ QgsAmbientOcclusionSettingsWidget::QgsAmbientOcclusionSettingsWidget( QWidget *p
+
+  mIntensitySpinBox->setClearValue( 0.5 );
+  mRadiusSpinBox->setClearValue( 25. );
+  mThresholdSpinBox->setClearValue( 0.5 );
```
This is an int spinbox, hence the failing build!
```suggestion
  mThresholdSpinBox->setClearValue( 50 );
```

wonder-sk replied:
argh, I had that fix in my copy but forgotten to push it last night :man_facepalming:

uclaros commented:
Should we not skip ssao when `depth == 1` to avoid darkening the background?

wonder-sk commented:
> Should we not skip ssao when depth == 1 to avoid darkening the background?

I think my preference would be to make also the background darker - the whole scene has then more consistent look in terms of brightness...

wonder-sk committed:
Clear button fix
